### PR TITLE
Fix wrong cfg80211->get_station implementation

### PR DESCRIPTION
**Pull request by qqgnoe466263**
Fix command failed by using "iw dev xxx link".

This is a failed demonstration:

Connected to aa:b7:12:fd:11:77 (on owl0)
        SSID: TestAP
        freq: 2437
command failed: No such file or directory (-2)

### CODE REVIEW

File: vwifi.c [hunk 1]
```diff
@@ -674,38 +674,36 @@ static int owl_get_station(struct wiphy *wiphy,
...
-                            BIT_ULL(NL80211_STA_INFO_RX_BYTES) |
-                            BIT_ULL(NL80211_STA_INFO_SIGNAL) |
-                            BIT_ULL(NL80211_STA_INFO_INACTIVE_TIME);
-
-            if (vif->sme_state == SME_CONNECTED) {
-                sinfo->filled |= BIT_ULL(NL80211_STA_INFO_CONNECTED_TIME);
-                sinfo->connected_time =
-                    jiffies_to_msecs(jiffies - vif->conn_time) / 1000;
-            }
+    if (memcmp(mac, vif->bssid, ETH_ALEN) != 0)
```
Comment (jserv) [inline vwifi.c:679]: Change the statement to
```c
if (!memcmp(mac, vif->bssid, ETH_ALEN))
```

File: vwifi.c [hunk 2]
```diff
@@ -674,38 +674,36 @@ static int owl_get_station(struct wiphy *wiphy,
...
+
+    sinfo->filled = BIT_ULL(NL80211_STA_INFO_TX_PACKETS) |
+                    BIT_ULL(NL80211_STA_INFO_RX_PACKETS) |
+                    BIT_ULL(NL80211_STA_INFO_TX_FAILED) |
+                    BIT_ULL(NL80211_STA_INFO_TX_BYTES) |
+                    BIT_ULL(NL80211_STA_INFO_RX_BYTES) |
+                    BIT_ULL(NL80211_STA_INFO_SIGNAL) |
+                    BIT_ULL(NL80211_STA_INFO_INACTIVE_TIME);
+
+    if (vif->sme_state == SME_CONNECTED) {
```
Comment (jserv) [inline vwifi.c:690]: Add some comments to make it informative.

**Comment by rickywu0421** [comment]
I have figured out the error comes from, and I think you have misunderstand the actual meaning of `get_station()`.

1. Don't be misleading by the name `get_station()`: It is to get the information of the station based of the parameter `const u8 *mac`, and in this semantic, the station can be every interface including station and AP (Yes, including AP), rather than get all stations running in STA mode which are connected to BSS. Following is the sample usage:
```
iw dev <The interface performing search> station get <MAC address of someone who you are interesting in>
```

2. I found that `iw dev <interface> link` will find the AP which was informed to kernel via passing the AP's MAC into `cfg80211_connect_result()`, and then perform `get_station()` to show the AP's information. The question is, the AP's MAC address is not the same as the interface running in AP mode, rather, is a random MAC address. So kernel just passes the "fake" MAC address  as `const u8 *Mac` into `get_station()`, and not surprisingly, fails to find the interface owning the "fake" MAC address.

In conclusion, you should modify `start_ap()` by inserting real MAC address into `ap_info_entry_t` rather than keep using the fake MAC address.

**Review by rickywu0421 — CHANGES_REQUESTED** [review]
After discussing with @qqgnoe466263, I figured out I was misunderstood the meaning of `get_station()` too. @qqgnoe466263  will explain the correct semantic of `get_station()` latter.

**Comment by qqgnoe466263** [comment]
Yes, you are right, I misunderstood get_station. And this problem can be fixed when I use real MAC to start AP. Thank you for your detailed explanation!

**Comment by jserv** [comment]
> Yes, you are right, I misunderstood get_station. And this problem can be fixed when I use real MAC to start AP. Thank you for your detailed explanation!

How about abandoning this pull request?

**Comment by qqgnoe466263** [comment]
I still discuss with @rickywu0421. And I'll check soon if this modification is not appropriate.

**Comment by rickywu0421** [comment]
> After thinking, i still feel that "get_station" can discuss and check. So I still discuss with @rickywu0421. And I'll check soon if this modification is not appropriate.

Yep, @qqgnoe466263 found that `get_station()` only do search when the parameter `const u8 mac` is the same as AP, I think it might be a security-based issue.

**Comment by rickywu0421** [comment]
> > Yes, you are right, I misunderstood get_station. And this problem can be fixed when I use real MAC to start AP. Thank you for your detailed explanation!
> 
> How about abandoning this pull request?

I think we could split this PR into two. One for changing `get_station()`, and another for fixing the error like title.

File: vwifi.c [hunk 3]
```diff
@@ -674,38 +674,36 @@ static int owl_get_station(struct wiphy *wiphy,
...
-            sinfo->rx_bytes = vif->stats.rx_bytes;
-            /* For CFG80211_SIGNAL_TYPE_MBM, value is expressed in dbm */
-            sinfo->signal = rand_int_smooth(-100, -30, jiffies);
-            sinfo->inactive_time = jiffies_to_msecs(jiffies - vif->active_time);
+    sinfo->tx_packets = vif->stats.tx_packets;
+    sinfo->rx_packets = vif->stats.rx_packets;
+    sinfo->tx_failed = vif->stats.tx_dropped;
+    sinfo->tx_bytes = vif->stats.tx_bytes;
+    sinfo->rx_bytes = vif->stats.rx_bytes;
+    /* For CFG80211_SIGNAL_TYPE_MBM, value is expressed in dbm */
```
Comment (jserv) [inline vwifi.c:701]: Should be "dBm"
See https://en.wikipedia.org/wiki/DBm

File: vwifi.c [hunk 4]
```diff
@@ -674,38 +674,36 @@ static int owl_get_station(struct wiphy *wiphy,
...
-                            BIT_ULL(NL80211_STA_INFO_RX_BYTES) |
-                            BIT_ULL(NL80211_STA_INFO_SIGNAL) |
-                            BIT_ULL(NL80211_STA_INFO_INACTIVE_TIME);
-
-            if (vif->sme_state == SME_CONNECTED) {
-                sinfo->filled |= BIT_ULL(NL80211_STA_INFO_CONNECTED_TIME);
-                sinfo->connected_time =
-                    jiffies_to_msecs(jiffies - vif->conn_time) / 1000;
-            }
+    if (!memcmp(mac, vif->bssid, ETH_ALEN))
```
Comment (rickywu0421) [inline vwifi.c:679]: Should this line be opposite case? (`if (memcmp(mac, vif->bssid, ETH_ALEN))`)

**Comment by jserv** [comment]
Thank @qqgnoe466263 for contributing!
